docs/3.x/extend/graphql.md

```diff
@@ -0,0 +1,1169 @@
+use mynamespace\helpers\Gql as MyGqlHelper;
+use craft\gql\base\GeneratorInterface;
+use craft\gql\GqlEntityRegistry;
+use craft\gql\TypeManager;
+
+class WidgetType implements GeneratorInterface
+{
+    public static function generateTypes($context = null): array
+    {
+        // Fetch all our pretend widget types to be used as contexts
+        $widgetTypes = Plugin::getInstance()->getAllWidgetTypes();
+        $gqlTypes = [];
+
+        foreach ($widgetTypes as $widgetType) {
+            // Get relevant scopes that may limit schema access
+            $requiredContexts = WidgetElement::gqlScopesByContext($widgetType);
+
+            // Ignore this widget variation if the schema doesn’t include it
+            if (!MyGqlHelper::isSchemaAwareOf($requiredContexts)) {
+                continue;
+            }
+
+            // Generate a GQL type for this widget type
+            $type = static::generateType($widgetType);
+            $gqlTypes[$type->name] = $type;
+        }
+
+        return $gqlTypes;
+    }
+
+    public static function generateType($context): ObjectType
+    {
+        // Get the intended GQL type name as determined by the element type
+        $typeName = WidgetElement::gqlTypeNameByContext($widgetType);
+
+        // Get element’s user-defined content fields and
+        $contentFieldGqlTypes = self::getContentFields($context);
+
+        // Merge in GQL types for the widget element’s own custom fields
+        $widgetFields = TypeManager::prepareFieldDefinitions(
+            array_merge(
+                WidgetInterface::getFieldDefinitions(),
+                $contentFieldGqlTypes
+            ),
+            $typeName
+        );
+
+        // Return the type if it exists, otherwise create and return it
+        return GqlEntityRegistry::getEntity($typeName) ?:
+            GqlEntityRegistry::createEntity(
+                $typeName,
+                new WidgetType([
+                    'name' => $typeName,
+                    'fields' => function() use ($widgetFields) {
+                        return $widgetFields;
+                    },
+                ])
+            );
+    }
+}
+```
+
+The term “context” here is deliberately vague because you’re the one that decides what that should be.
+
+Since Craft elements and field types are integral to the GraphQL API, their base classes include methods for describing their type names:
+
+- <craft3:craft\base\Element::getGqlTypeName()>
+- <craft3:craft\base\Element::gqlTypeNameByContext()>
+- <craft3:craft\base\Field::getContentGqlType()>
+
+These include sensible defaults since Craft can make some assumptions about how elements and fields will be used.
+
+::: tip
+A field can also use <craft3:craft\base\Field::includeInGqlSchema()>—`true` by default—to determine whether it should appear in a given schema.
+:::
+
+## Directives
+
+Directives return types that can be used to transform result data in specified locations relative to the GraphQL query.
```

Comment:
Maybe worth stressing the point that directives are applied to returned results. That is - they are invoked once Craft has returned all it will return for that query and, as such, it's not possible to use directives to influence what Craft returns. It's only possible to massage the results Craft has already returned.